packages/x-data-grid-pro/src/hooks/features/serverSideData/interfaces.ts

```diff
@@ -0,0 +1,10 @@
+import { GridRowId } from '@mui/x-data-grid';
+
+export interface GridServerSideDataInternalCache {
+  groupKeys: any[];
+}
+
+export interface GridServerSideDataState {
+  loading: Record<GridRowId, boolean>;
+  errors: Record<GridRowId, any>;
```

Comment:
So this means a row could have an error and be loading at the same time?

Otherwise if it's an either/or situation, it might be worth to reduce to a single state field like `status: Record<id, Loading | Error | Loaded>`.

Reply:
It's not currently the case but we might need to preserve the error for a loading row. Moreover, since we store actual error message in `state.errors`, it might not play well with a single status field.

For example, if we use an enum like:

```tsx
enum GridRowFetchStatus {
  Loading,
  Error,
  Success,
}
```

For Loading and Success, we can use `if (status === GridRowFetchStatus.Loading) {` but for preserving error message, we will still have to introduce a new state field such as `state.errors`. I propose we keep it like this until a specific need arises. Wdyt?

Reply:
This only makes sense if we don't store error messages for loading rows, but we can use something like algebraic data types:

```typescript
class StatusBase {}

class Loading extends StatusBase {}
class Error extends StatusBase {
  message: string
}
class Success<T> extends StatusBase {
  data: T
}

type Status = Loading | Error | Success<GridRow>
```

Which provides 100% type safety, making it impossible to access data or error messages when there's none. This might be overkill though, I'm not insisting on this.

> we might need to preserve the error for a loading row

I'd be mindful of YAGNI when designing the state. Restricting the scope to what we're really going to implement can help manage complexity.